### PR TITLE
add methods to config

### PR DESCRIPTION
**Pull request by cssivision**
a simple pr, lack of many things, we can discuss API base on this.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -204,6 +210,32 @@ impl Config {
...
+                ScopedIp::from(IpAddr::V6(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1))),
+            ]
+        } else {
+            self.nameservers.to_vec()
+        }
+    }
+
+    /// Get domain or hostname
+    pub fn get_domain_or_hostname(&self) -> Option<String> {
+        self.domain.clone().or(get_hostname())
```
Comment (tailhook) [inline src/config.rs]: Domain doesn't fallback to a hostname, but:
> the domain is determined from the local hostname returned by  gethostname(2);  the domain  part is taken to be everything after the first '.'.

File: src/config.rs [hunk 2]
```diff
@@ -204,6 +210,32 @@ impl Config {
...
+                ScopedIp::from(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1))), 
+                ScopedIp::from(IpAddr::V6(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1))),
+            ]
+        } else {
+            self.nameservers.to_vec()
+        }
+    }
+
+    /// Get domain or hostname
+    pub fn get_domain_or_hostname(&self) -> Option<String> {
```
Comment (tailhook) [inline src/config.rs]: I'm not sure but I think I would do `get_domain_from_env`, and that function would look env variable first and than these two things. The idea is that `env` used here is a broad sense. Although may be there is a better name?

Or do you think there is a use case for getting domain without looking at env variables?

File: src/config.rs [hunk 3]
```diff
@@ -204,6 +210,32 @@ impl Config {
...
+    pub fn glibc_normalize(&mut self) {
+        self.nameservers.truncate(NAMESERVER_LIMIT);
+        self.search = self.search.take().map(|mut s| {
+            s.truncate(SEARCH_LIMIT); 
+            s
+        });
+    }
+
+    /// Get nameserver or on the local machine
+    pub fn get_nameserver_or_local(&self) -> Vec<ScopedIp> {
```
Comment (tailhook) [inline src/config.rs]: nameserver**S** ?

File: src/config.rs [hunk 4]
```diff
@@ -204,6 +210,32 @@ impl Config {
         self.search = Some(search);
         self.last_search = LastSearch::Search;
     }
+
+    /// Normalize config.
```
Comment (tailhook) [inline src/config.rs]: The docstring should document current behavior more specifically.

Also I'm thinking by emitting warnings somehow, still don't want to depend on logging. So let's keep bare normalization for now, and will add another method with warnings in future.

File: src/config.rs [hunk 5]
```diff
@@ -1,6 +1,12 @@
 use std::iter::{IntoIterator, Iterator};
 use std::slice::Iter;
 use {grammar, Network, ParseError, ScopedIp};
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+
+use hostname::get_hostname;
```
Comment (tailhook) [inline src/config.rs]: Let's put it under a feature gate. I like that the parser could be platform independent, i.e. one may want to compile it to wasm and put an online validator for resolv.conf file.

File: src/config.rs (hunk 2 above)
Comment (cssivision) [reply src/config.rs]: you mean `LOCALDOMAIN`? 

this env varibale is for `search` keyword.
> The search keyword of a system's resolv.conf file can be overridden
       on a per-process basis by setting the environment variable
       LOCALDOMAIN to a space-separated list of search domains.

File: src/config.rs (hunk 5 above)
Comment (cssivision) [reply src/config.rs]: feature name? `unix`?

Comment (cssivision) [reply src/config.rs]: maybe other methods on config should be put in this feature? like `glibc_normalize`, `get_nameservers_or_local` which both for linux.

File: src/config.rs (hunk 2 above)
Comment (tailhook) [reply src/config.rs]: Okay, I was confused by env name :)

Okay, still not sure about function name. Something like `compute_domain` maybe.

File: src/config.rs (hunk 5 above)
Comment (tailhook) [reply src/config.rs]: I believe `hostname` crate works on windows/OS X too. Maybe `os` or `system` ?

Comment (tailhook) [reply src/config.rs]: And things that don't rely on platform API's shouldn't be feature-gated. I.e. there are use cases to validate linux'ish resolv.conf on OS X or Windows (for example to prepare a docker container).

So `glibc_normalize` and `get_nameservers_or_local` should be kept without feature gate.

File: src/config.rs (hunk 2 above)
Comment (tailhook) [reply src/config.rs]: Or `get_system_domain` ? (especially if we name feature gate `system`)

File: tests/lib.rs [hunk 6]
```diff
@@ -222,3 +222,67 @@ fn test_parse_macos_conf() {
...
+            ScopedIp::from(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1))), 
+            ScopedIp::from(IpAddr::V6(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1))),
+        ],
+        config.get_nameservers_or_local()
+    );
+}
+
+#[test]
+#[cfg(feature = "system")]
+#[ignore]
```
Comment (cssivision) [inline tests/lib.rs]: different value in different machine, so ignore it.

File: src/config.rs [hunk 7]
```diff
@@ -204,6 +211,40 @@ impl Config {
...
+    }
+
+    /// Get domain if not exists return hostname.
+    #[cfg(feature = "system")]
+    pub fn get_system_domain(&self) -> Option<String> {
+        let hn = get_hostname().map(|s| {
+            if let Some(pos) = s.find('.') {
+                return s[pos + 1..].to_string();
+            };
+            s
```
Comment (tailhook) [inline src/config.rs]: > the domain part is taken  to  be everything after the first '.'.  Finally, if the hostname does not contain a domain part, the root domain is assumed.

So it's either empty string here or None, but not host name. I think None is better.

File: src/config.rs [hunk 8]
```diff
@@ -204,6 +211,40 @@ impl Config {
...
+    #[cfg(feature = "system")]
+    pub fn get_system_domain(&self) -> Option<String> {
+        let hn = get_hostname().map(|s| {
+            if let Some(pos) = s.find('.') {
+                return s[pos + 1..].to_string();
+            };
+            s 
+        });
+
+        self.domain.clone().or(hn)
```
Comment (tailhook) [inline src/config.rs]: It's better to skip `get_hostname()` call if domain is already set.

File: tests/lib.rs [hunk 9]
```diff
@@ -222,3 +222,67 @@ fn test_parse_macos_conf() {
...
+                None,
+            ),
+            ScopedIp::V6(
+                Ipv6Addr::new(0x2001, 0x4860, 0x4860, 0, 0, 0, 0, 0x8844),
+                None,
+            ),
+            ScopedIp::V4(Ipv4Addr::new(8, 8, 8, 8)),
+        ],
+        config.nameservers
+    );
```
Comment (tailhook) [inline tests/lib.rs]: why `search` is unchecked?

Comment (cssivision) [reply tests/lib.rs]: fixed

**Comment by tailhook** [comment]
Looks good. Merged. Thanks!

**Pull request merged**